middle_end/flambda2/terms/flambda_primitive.ml

```diff
@@ -1056,6 +1061,7 @@ let effects_and_coeffects_of_unary_primitive p : Effects_and_coeffects.t =
     ( Only_generative_effects Mutable (* Mutable is conservative *),
       Has_coeffects,
       Strict )
+  | Get_header -> No_effects, No_coeffects, Strict
```

Comment:
Here as well as for the cse, I'm a bit worried about the colors bits in the headers, since these are in effect mutable by the GC, and whose changes could reasonably be expected to be observable by users of `get_header`. However, sharing introduced by the cse, or some operations being moved around[1] (because `get_header` has no effects/coeffects) effectively make the color bits of the header not usable, which could be fine, but the tests added later make use of the color bits.

[1]: for instance a `get_header` could be moved across some calls to GC-related functions (such as `Gc.major`, `Gc.major_slice`), which would break some code that tests color bits before and after such functions, unless I missed something.

Reply:
We can give `get_header` coeffects to avoid these problems, is that right?

Reply:
It will help, since it will prevent that primitive from being commuted with a function that **explicitly** calls the GC (e.g. `Gc.major`, etc...)

However, `to_cmm` would still be allowed to move allocations (i.e. expressions with only immutable generative effects) across it (which is necessary to allow `to_cmm` to sink down some allocations), so in particular the result of such a primitive could change between classic and non-classic mode potentially, but I don't think it's that big of a problem (i.e. in a program with allocations everywhere, the colors bits are already more or less random).

Reply:
True, in that case maybe we shouldn't give it coeffects.  We can add a comment noting that the GC bits will be unreliable.

Reply:
Adding coeffects would seem to just change how unreliable the colour flags are, so I think it's ok to leave this as pure.